DRA in Kubernetes 1.30: adds structured parameters (#45287)
kubernetes/website: content/en/docs/concepts/scheduling-eviction/dynamic-resource-allocation.md

<!-- overview -->

{{< feature-state for_k8s_version="v1.26" state="alpha" >}}

> {{< feature-state for_k8s_version="v1.30" state="alpha" >}}

Author: I'm not sure what to document here: the first time this became available (1.26), the current status, or both?

If we include the current status, we have to refresh the version on each release. This had not been done for 1.28 and 1.29 because there were no other changes.

I had bumped the version from 1.26 to 1.27 because of the API change, but I am no longer sure whether that was the right thing to do, so I am going back to 1.26 here.

Reviewer: We show readers whether the feature is alpha, beta, GA (or deprecated), and the version when the last transition occurred.

Provided you're targeting a recent-enough _dev-1.30_, you can do this:
```markdown
{{< feature-state feature_gate_name="DynamicResourceAllocation" >}}
```

and make sure that the feature gate file [`content/en/docs/reference/command-line-tools-reference/feature-gates/dynamic-resource-allocation.md`](https://github.com/kubernetes/website/blob/main/content/en/docs/reference/command-line-tools-reference/feature-gates/dynamic-resource-allocation.md) is current (I don't think there'd be any change needed).

Dynamic resource allocation is an API for requesting and sharing resources
between pods and containers inside a pod. It is a generalization of the
persistent volumes API for generic resources. Third-party resource drivers are
responsible for tracking and allocating resources, with additional support
provided by Kubernetes via "structured parameters" (introduced in Kubernetes 1.30).

> Different kinds of resources support arbitrary parameters for defining requirements and

Reviewer 2: I think we should expand on the distinction here a bit more in the intro.

Author: I added one sentence:
> When a driver uses structured parameters, Kubernetes handles scheduling
> and resource allocation without having to communicate with the driver.

<!-- overview -->

{{< feature-state feature_gate_name="DynamicResourceAllocation" >}}

Dynamic resource allocation is an API for requesting and sharing resources
between pods and containers inside a pod. It is a generalization of the
persistent volumes API for generic resources. Third-party resource drivers are
responsible for tracking and allocating resources, with additional support

> provided by Kubernetes via "structured parameters" (introduced in Kubernetes 1.30).

Reviewer: ```suggestion
provided by Kubernetes via _structured parameters_ (introduced in Kubernetes 1.30).
```

Please also consider adding “structured parameter” into our glossary.

Author: I applied your suggestion and squashed (forgot that earlier).

I'll keep the glossary in mind for the next update. Right now I feel it's too early - we still need to find out whether the name and concept survives.

  to coordinate pod scheduling when ResourceClaims need to be allocated
  for a Pod.

ResourceSlice
: Used with structured parameters to publish information about resources
  that are available in the cluster.

ResourceClaimParameters
: Contain the parameters for a ResourceClaim which influence scheduling,
  in a format that is understood by Kubernetes (the "structured parameter

>   model"). Initialization parameters may be embedded in an opaque
>   extension, for use by the vendor driver when setting up the underlying
>   resource.

Reviewer 2: Let's just say "additional" instead of "initialization" sicne we don't atually know exactly what these vendor-specific parameters will be used for.

```suggestion
  model"). Additional parameters may be embedded in an opaque
  extension, for use by the vendor driver when setting up the underlying
  resource.
```

Author: It was called "initialization" parameters above already. I think in the KEP we used "configuration". But "additional" works best here.

ResourceClassParameters
: Similar to ResourceClaimParameters, the ResourceClassParameters provides
  a type for ResourceClass parameters which is understood by Kubernetes.

Parameters for ResourceClass and ResourceClaim are stored in separate objects,
typically using the type defined by a {{< glossary_tooltip
term_id="CustomResourceDefinition" text="CRD" >}} that was created when
installing a resource driver.

The developer of a resource driver decides whether they want to handle those
parameters in their own controller or want to use structured parameters. A
custom controller provides more flexibility, but cluster autoscaling is not
going to work reliably for node-local resources. Structured parameters enable

> cluster autoscaling, but might not satisfy all use-cases.

Author: s/kubernetes/Kubernetes/, otherwise fine with me.

> When a driver uses structured parameters, it is still possible to let the
> end-user specify parameters with CRDs. The driver then needs to translate those
> custom parameters into the in-tree types. Alternatively, a driver may also
> document how to use the in-tree types directly.

Reviewer 2: ```suggestion
When a driver uses structured parameters, it is still possible to let the
end-user specify parameters with vendor-specific CRDs. When doing so, the driver needs to translate those
custom parameters into the in-tree types. Alternatively, a driver may also
document how to use the in-tree types directly.
```

{{< /note >}}

### With structured parameters

> When a driver uses structured parameters, the scheduler takes over the
> responsibility of allocating a ResourceClaim once a pod needs the
> ResourceClaim. It does that by retrieving the information about available
> resources from ResourceSlice objects, tracking which of those are in use based
> on already allocated ResourceClaims, and then selecting resources which match
> the requirements defined by ResourceClaimParameters and (if specified)
> ResourceClassParameters and which are available on a potential node.

Reviewer 2: ```suggestion
When a driver uses structured parameters, the scheduler takes over the
responsibility of allocating resources to a ResourceClaim whenever a pod needs them. It does so by retrieving the full list of available
resources from ResourceSlice objects, tracking which of those resources have already been allocated to existing ResourceClaims, and then selecting from those resources that remain.  The exact resources selected are subject to the constraints provided in any ResourceClaimParameters or
ResourceClassParameters associated with the claim.
```

Author: s/claim/ResourceClaim/, for consistency.

> The chosen resource is recorded in the ResourceClaim status together with the
> initialization parameters, so when a pod is about to start on a node, the
> resource driver on the node has all information that it needs to prepare the
> resource.

Reviewer 2: Do we want to say initialization parameters or vendor-specific parameters?

```suggestion
The chosen resource is recorded in the ResourceClaim status together with any vendor-specific parameters, so when a pod is about to start on a node, the
resource driver on the node has all the information it needs to prepare the
resource.
```
